spec/wibox/layout/align_spec.lua

```diff
@@ -223,6 +223,190 @@ describe("wibox.layout.align", function()
+
+                    it("layout", function()
+                        assert.widget_layout(layout, { 5, 100 }, {
+                            p(first,  0,  0, 5, 20),
+                            p(third,  0, 80, 5, 20),
+                            p(second, 0, 20, 5, 60),
+                        })
+                    end)
+                end)
+
+                describe("without enough width", function()
+                    it("fit", function()
+                        assert.widget_fit(layout, { 100, 30 }, { 15, 30 })
+                    end)
+
+                    it("layout", function()
+                        assert.widget_layout(layout, { 100, 30 }, {
+                            p(first,  0,  0, 100, 20),
+                            p(third,  0, 20, 100, 10),
+                        })
+                    end)
+                end)
+            end)
+
+            describe("with third larger", function()
+                local first, second, third
+
+                before_each(function()
+                    first = utils.widget_stub(10, 10)
+                    second = utils.widget_stub(15, 15)
+                    third = utils.widget_stub(10, 20)
+
+                    layout:set_first(first)
+                    layout:set_second(second)
+                    layout:set_third(third)
+                end)
+
+                describe("with enough space", function()
+                    it("fit", function()
+                        assert.widget_fit(layout, { 100, 100 }, { 15, 45 })
+                    end)
+
+                    it("layout", function()
+                        assert.widget_layout(layout, { 100, 100 }, {
+                            p(first,  0,  0, 100, 20),
+                            p(third,  0, 80, 100, 20),
+                            p(second, 0, 20, 100, 60),
+                        })
+                    end)
+                end)
+
+                describe("without enough height", function()
+                    it("fit", function()
+                        assert.widget_fit(layout, { 5, 100 }, { 5, 45 })
+                    end)
+
+                    it("layout", function()
+                        assert.widget_layout(layout, { 5, 100 }, {
+                            p(first,  0,  0, 5, 20),
+                            p(third,  0, 80, 5, 20),
+                            p(second, 0, 20, 5, 60),
+                        })
+                    end)
+                end)
+
+                describe("without enough width", function()
+                    it("fit", function()
+                        assert.widget_fit(layout, { 100, 30 }, { 15, 30 })
+                    end)
+
+                    it("layout", function()
+                        assert.widget_layout(layout, { 100, 30 }, {
+                            p(third,  0, 10, 100, 20),
+                            p(first,  0, 0, 100, 10),
+                        })
+                    end)
+                end)
```

Comment:
Same extra test cases as above. Probably even more important here, to make sure things really aren't placed "first come, first serve".

Reply:
OK, I will add more test cases.

Reply:
I replaced this 1 unit test for not enough width by 6 for different width configurations for each of the 3 size cases in [6499cc4](https://github.com/awesomeWM/awesome/pull/3588/commits/6499cc496ae26b3aac941badebcb81ec0f4d5f98).
Now, in total, for the width, we have 7 tests:
* More that required for all 3 widgets (expand the smallest side and the middle)
* Just enough to fit 3 widgets (stop expanding the smallest side and the middle, place widgets right one after another)
* Barely enough to fit 3 widgets (shrink center widget)
* Just enough to fit 2 widgets (place only side widgets right one after another)
* Barely enough to fit 2 widgets (shrink the smallest side)
* Just enough to fit 1 widget (place only the longest side)
* Barely enough to fit 1 widget (place only a part of the longest side)